### PR TITLE
led tool: toggle fault LEDs

### PR DESCRIPTION
**Pull request by SunnySrivastava1984**
The commit implements base framework for led tool with an option to toggle fault LEDs based on the functional status of the FRU.

The tool will be auto triggered with different optiosn on BMC boot and system power on, driven via system-d service files.

Tool can be used via command line as well.

### CODE REVIEW

File: tools/toggle-fault-leds.hpp [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+            {
+                // Asserted is inverse of functional status.
+                utility::setProperty<bool>(
+                    "xyz.openbmc_project.LED.GroupManager", path,
+                    "xyz.openbmc_project.Led.Group", "Asserted", !isFunctional);
+            }
+        }
+        else
+        {
+            std::cout << "No Endpoints for path: " << objectPath << std::endl;
```
Comment (branupama) [inline tools/toggle-fault-leds.hpp:99]: Nits: endpoints

File: tools/utility.hpp [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+
+namespace utility
+{
+
+/**
+ * @brief An API to make GetSubTree mapper call.
+ *
+ * @return MapperResponse - Response from the API.
+ */
+// map<Inventory_path, map<service_name, vector<interfaces>>>
```
Comment (branupama) [inline tools/utility.hpp:18]: Nits: inventory_path , interface ?

File: tools/utility.hpp [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+using MapperResponse =
+    std::map<std::string, std::map<std::string, std::vector<std::string>>>;
+MapperResponse
```
Comment (branupama) [inline tools/utility.hpp:21]: new line ?

File: tools/utility.hpp [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+    }
+    return result;
+}
+
+/**
+ * @brief Templated API to set D-Bus property.
+ *
+ * @param[in] serviceName - D-Bus service name hosting the object path.
+ * @param[in] objectPath - D-Bus object path hosting the interface.
+ * @param[in] interface - D-Bus interface hosting the porperty.
```
Comment (branupama) [inline tools/utility.hpp:55]: Nits: property

File: tools/utility.hpp [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+                  << " with error: " << e.what() << std::endl;
+    }
+}
+
+/**
+ * @brief Templated API to get D-Bus property.
+ *
+ * @param[in] serviceName - D-Bus service name hosting the object path.
+ * @param[in] objectPath - D-Bus object path hosting the interface.
+ * @param[in] interface - D-Bus interface hosting the porperty.
```
Comment (branupama) [inline tools/utility.hpp:89]: Nits: property

File: tools/utility.hpp [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+                  << " with error: " << e.what() << std::endl;
+    }
+
+    return result;
+}
+
+/**
+ * @brief An API to check chassis power state.
+ *
+ * @return bool - Power state of chassis, false otherwise.
```
Comment (branupama) [inline tools/utility.hpp:125]: when true will be set is not specified.
so "Power state of chassis" is enough ?

**Comment by branupama** [comment]
spell mistake in commit message.

File: tools/toggle-fault-leds.hpp [hunk 7]
```diff
@@ -0,0 +1,102 @@
+#pragma once
+#include "utility.hpp"
+
+/**
+ * @brief API to toggle fault LEDs according to the given functional status.
+ *
+ * The API sets "Functional" property hosted by operational status interface for
+ * all the FRUs under root path "/xyz/openbmc_project/inventory". It also
+ * fetches associated LEDs object path for FRUs, hosted under LED manager and
+ * sets asserted vlaue for accordingly.
```
[the remark this answers is not in the document]
Comment (SunnySrivastava1984) [reply tools/toggle-fault-leds.hpp:10]: Done.

File: tools/toggle-fault-leds.hpp [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+    if (subTree.size() == 0)
+    {
+        std::cout
+            << "No sub tree found for OperationalStatus interfaces. Exiting."
+            << std::endl;
+
+        return;
+    }
+
+    for (const auto& [objectPath, serviceInterfceMap] : subTree)
```
[the remark this answers is not in the document]
Comment (SunnySrivastava1984) [reply tools/toggle-fault-leds.hpp:55]: Done.

File: tools/toggle-fault-leds.hpp (hunk 1 above)
Comment (SunnySrivastava1984) [reply tools/toggle-fault-leds.hpp:99]: Done.

File: tools/utility.hpp (hunk 2 above)
Comment (SunnySrivastava1984) [reply tools/utility.hpp:18]: Done.

File: tools/utility.hpp (hunk 3 above)
Comment (SunnySrivastava1984) [reply tools/utility.hpp:21]: Done.

File: tools/utility.hpp (hunk 4 above)
Comment (SunnySrivastava1984) [reply tools/utility.hpp:55]: Done.

File: tools/utility.hpp (hunk 5 above)
Comment (SunnySrivastava1984) [reply tools/utility.hpp:89]: Done.

File: tools/utility.hpp (hunk 6 above)
Comment (SunnySrivastava1984) [reply tools/utility.hpp:125]: Added extra info.

File: tools/toggle-fault-leds.hpp [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+
+    for (const auto& [objectPath, serviceInterfaceMap] : subTree)
+    {
+        if (std::any_of(skipOperationalStatusForFRUs.cbegin(),
+                        skipOperationalStatusForFRUs.cend(),
+                        [&objectPath](const std::string& str) {
+            return objectPath.find(str) != std::string::npos;
+        }))
+        {
+            //"Skip setting operational status for these paths."
```
Comment (branupama) [inline tools/toggle-fault-leds.hpp:63]: Nits: do we need " here ?

File: tools/toggle-fault-leds.hpp [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+        }
+
+        utility::setProperty<bool>(
+            "xyz.openbmc_project.Inventory.Manager", objectPath,
+            "xyz.openbmc_project.State.Decorator.OperationalStatus",
+            "Functional", isFunctional);
+
+        if (std::any_of(frusWithoutLED.cbegin(), frusWithoutLED.cend(),
+                        [&objectPath](const std::string& str) {
+            return objectPath.find(str) != std::string::npos;
```
Comment (branupama) [inline tools/toggle-fault-leds.hpp:74]: is 'aFru' better instead of str ?

File: tools/toggle-fault-leds.hpp [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+    std::vector<std::string> frusWithoutLED{
+        "pcie_card", "usb",    "drive",       "ethernet",  "fan0_",
+        "fan1_",     "fan2_",  "fan3_",       "fan4_",     "fan5_",
+        "rdx",       "cables", "displayport", "pcieslot12"};
+
+    if (utility::isChassisOn())
+    {
+        std::cout << "Abort toggling fault LED. Chassis is in power on state."
+                  << std::endl;
+
```
Comment (branupama) [inline tools/toggle-fault-leds.hpp:36]: is empty line required ?

File: tools/toggle-fault-leds.hpp [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+    utility::MapperResponse subTree = utility::getObjectSubtreeForInterfaces(
+        "/xyz/openbmc_project/inventory", 0, interfaces);
+
+    if (subTree.size() == 0)
+    {
+        std::cout
+            << "No sub tree found for OperationalStatus interfaces. Exiting."
+            << std::endl;
+
```
Comment (branupama) [inline tools/toggle-fault-leds.hpp:51]: Empty line required ?

File: tools/toggle-fault-leds.hpp (hunk 11 above)
Comment (SunnySrivastava1984) [reply tools/toggle-fault-leds.hpp:36]: Requirement can't be defined here. It makes the code looks clean and readable.

File: tools/toggle-fault-leds.hpp (hunk 12 above)
Comment (SunnySrivastava1984) [reply tools/toggle-fault-leds.hpp:51]: Requirement can't be defined here. It makes the code looks clean and readable.

File: tools/toggle-fault-leds.hpp [hunk 13]
```diff
@@ -0,0 +1,102 @@
+#pragma once
+#include "utility.hpp"
+
+/**
+ * @brief API to toggle fault LEDs according to the given functional status.
+ *
+ * The API sets "Functional" property hosted by operational status interface for
+ * all the FRUs under root path "/xyz/openbmc_project/inventory". It also
+ * fetches associated LEDs object path for FRUs, hosted under LED manager and
+ * sets asserted value for accordingly.
```
[the remark this answers is not in the document]
Comment (SunnySrivastava1984) [reply tools/toggle-fault-leds.hpp]: Done.

File: tools/toggle-fault-leds.hpp (hunk 9 above)
Comment (SunnySrivastava1984) [reply tools/toggle-fault-leds.hpp]: Done.

File: tools/toggle-fault-leds.hpp (hunk 10 above)
Comment (SunnySrivastava1984) [reply tools/toggle-fault-leds.hpp]: Done.

**Review by branupama — APPROVED** [review]

**Review by jinuthomas — APPROVED** [review]

**Pull request merged**